### PR TITLE
improving performance of PuppiProducer

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR is an attempt to improve the performance (in terms of execution time) of the `PuppiProducer` plugin.

No changes in the outputs are expected.

Most of these technical changes are not in the producer itself, but in the `PuppiContainer` class; all changes are related (directly or indirectly) to the function `var_within_R` (the "alpha" variable on which the PUPPI weight is based), which appears to be the most time-consuming part of the algorithm.

The table below summarizes a few timing estimates for the `puppi` module in the RECO step for a Run-3 and Phase-2 workflow.

| wf | time pre-PR [ms] | time post-PR [ms] |
|---:|---:|---:|
| Run-3 RECO [(QCD MC)](https://cmsweb.cern.ch/das/request?input=dataset%3D%2FRelValQCD_FlatPt_15_3000HS_14%2FCMSSW_11_2_0_pre3-PU_112X_mcRun3_2021_realistic_v5-v1%2FGEN-SIM-DIGI-RAW&instance=prod/global)    |  74.2 |  33.2 |
| Phase-2 RECO [(QCD MC)](https://cmsweb.cern.ch/das/request?input=dataset%3D%2FRelValQCD_Pt15To7000_Flat_14%2FCMSSW_11_2_0_pre3-PU25ns_110X_mcRun4_realistic_v3_2026D49PU200_rsb-v1%2FGEN-SIM-DIGI-RAW&instance=prod/global) |  620.1 | 255.9  |

The numbers are taken from the `FastTimerService`, running on 100 events; they suggest that the overall reduction in timing is approx. 55-60%.

Here is a short summary of the changes in the PR (below, `goodVar` and `var_within_R` are used interchangeably, since the former function is simply a call to the latter):

* `PuppiCandidate` is converted to a simple `struct`, removing its inheritance from `fastjet::PseudoJet` (looking at the PUPPI implementation as a whole, the usage of `fastjet` does not seem necessary); this change accounts for most of the speedup (approx. 35-40% speedup);

* skip the calculation of `goodVar` when not needed, e.g. for candidates that are assigned a default weight (0, or 1); this seems to bring roughly a 10-15% improvement in performance.

* another bit of speedup is obtained by taking one condition (`puppi_id != 3`) out of `var_within_R`, and using it just once in `PuppiContainer::initialize` to fill the collections later used as inputs to `goodVar`; this leads to adding one more data member (a `vector` of `PuppiCandidate`s) to the `PuppiContainer` class.

In this PR, there are two spots where things could certainly be improved (but improving them will lead to small changes in the outputs):

* in `PuppiContainer::initialize`, an instance of `fastjet::PseudoJet` ([here](https://github.com/cms-sw/cmssw/compare/master...missirol:devel_112X_puppi_speedup2?expand=1#diff-920ad87cd21790b241508fda9a732044R49)) is used to obtain the 4-momenta of the candidates used by PUPPI; this was the only way I could find to maintain the same numbers as pre-PR (this may be due to the fact that the inputs used for `reset_PtYPhiM` are floats, and then `fastjet` recomputes and stores the p4 using doubles); one way to simplify this (assuming the differences in the outputs are ultimately not significant) would be to remove `PuppiCandidate` altogether and simply use the `RecoObj` class already in use in the `PuppiProducer` (maybe, changing the floats in `RecoObj` to doubles).

* in `goodVar`, the `DeltaR2` between two given candidates is calculated (at most) twice ([here](https://github.com/cms-sw/cmssw/compare/master...missirol:devel_112X_puppi_speedup2?expand=1#diff-920ad87cd21790b241508fda9a732044R96)), first with rapidity, then with pseudo-rapidity; this is not modified by this PR (otherwise, the outputs would inevitably change), but maybe it’s something that could be simplified.

FYI: @ahinzmann @lathomas @kirschen

#### PR validation:

Checked that the PUPPI weights of all PF candidates are unchanged for 200 events (100 for a Run-3 workflow, and 100 for a Phase-2 workflow), for both `puppi` (used for jets) and `puppiNoLep` (used for MET).
Standard workflows, i.e. `runTheMatrix.py -l limited -i all --ibeos`, ran successfully.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

### CODE REVIEW

**Comment by slava77** [comment]
@mrodozov @smuzaffar 
do we have some problems with webhooks?
this one was posted for 14 minutes already (at the time of my message)

**Comment by mrodozov** [comment]
@slava77 the remote calls to jenkins hanged with proxy error 502 and I'm trying to restart the service but I don't have sudo rights to do it

**Comment by davidlange6** [comment]
It seems there are other problems at cern today (at least affecting cms computing infrastructure)

On Aug 18, 2020, at 3:33 PM, Mircho Rodozov <notifications@github.com<mailto:notifications@github.com>> wrote:



@slava77<https://github.com/slava77> the remote calls to jenkins hanged with proxy error 502 and I'm trying to restart the service but I don't have sudo rights to do it

**Comment by mrodozov** [comment]
@davidlange6 do you by any chance happen to have root permissions for the jenkins03 machine ?

**Comment by davidlange6** [comment]
No, I don’t - perhaps ask IT to reboot it?

On Aug 18, 2020, at 4:45 PM, Mircho Rodozov <notifications@github.com<mailto:notifications@github.com>> wrote:



@davidlange6<https://github.com/davidlange6> do you by any chance happen to have root permissions for the jenkins03 machine ?

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 1]
```diff
@@ -77,70 +86,51 @@ double PuppiContainer::var_within_R(int iId,
...
-    return 1;
+    return 1.;
 
-  //this is a circle in rapidity-phi
-  //it would make more sense to have var definition consistent
-  //fastjet::Selector sel = fastjet::SelectorCircle(R);
-  //sel.set_reference(centre);
-  //the original code used Selector infrastructure: it is too heavy here
-  //logic of SelectorCircle is preserved below
+  double const r2(R * R);
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc:91]: I guess in some sense it's matter of style.
Still, what's wrong with ` = ` ?

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 2]
```diff
@@ -77,70 +86,51 @@ double PuppiContainer::var_within_R(int iId,
...
-      continue;
-    //squared_distance is in (y,phi) coords: rap() has faster access -> check it first
-    if (std::abs(part.rap() - centre.rap()) < R && part.squared_distance(centre) < r2) {
-      near_dR2s.push_back(reco::deltaR2(part, centre));
-      near_pts.push_back(part.pt());
+  for (auto const &cand : particles) {
+    if (std::abs(cand.rapidity - centre.rapidity) < R) {
+      auto const dr2y(reco::deltaR2(cand.rapidity, cand.phi, centre.rapidity, centre.phi));
+      if (dr2y < r2) {
+        auto const dr2(reco::deltaR2(cand.eta, cand.phi, centre.eta, centre.phi));
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc:98]: it looks like only ```1./dr2``` is used in a few places below, some speedup should be possible from computing this inversion here

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 3]
```diff
@@ -150,66 +140,69 @@ void PuppiContainer::getRMSAvg(int iOpt,
...
-    if (!pCharged)
-      pVal = goodVar(iConstits[i0], iParticles, pAlgo, pCone);
-    if (pCharged)
-      pVal = goodVar(iConstits[i0], iChargedParticles, pAlgo, pCone);
+    double pVal = -1;
+    // calculate goodVar only for candidates that (1) will not be assigned a predefined weight (e.g 0, 1),
+    // or (2) are required for computations inside puppi-algos (see call to PuppiAlgo::add below)
+    if (((not(fApplyCHS and ((iConstits[i0].id == 1) or (iConstits[i0].id == 2)))) and (iConstits[i0].id != 3)) or
+        ((std::abs(iConstits[i0].eta) < fPuppiAlgo[pPupId].etaMaxExtrap()) and
+         ((iConstits[i0].id == 1) or (iConstits[i0].id == 2)))) {
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc:148]:
```suggestion
    bool pvRelated = iConstits[i0].id == 1 or iConstits[i0].id == 2;
    if (not(fApplyCHS and pvRelated) or
        (pvRelated and std::abs(iConstits[i0].eta) < fPuppiAlgo[pPupId].etaMaxExtrap()) ) {
```
(assuming I didn't get lost in parentheses) 
- IIUC, iConstits == 3 were already removed, so it's always true
- a common piece is precomputed
- somewhat excessive use of parentheses is reduced

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 4]
```diff
@@ -150,66 +140,69 @@ void PuppiContainer::getRMSAvg(int iOpt,
...
+      pVal = goodVar(iConstits[i0], pCharged ? iChargedParticles : iParticles, pAlgo, pCone);
+    }
     fVals.push_back(pVal);
-    //if(std::isnan(pVal) || std::isinf(pVal)) cerr << "====> Value is Nan " << pVal << " == " << iConstits[i0].pt() << " -- " << iConstits[i0].eta() << endl;
+    //if(std::isnan(pVal) || std::isinf(pVal)) cerr << "====> Value is Nan " << pVal << " == " << iConstits[i0].pt << " -- " << iConstits[i0].eta << endl;
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc:152]: can be dropped, apparently, it's a repeat of the code below

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 5]
```diff
@@ -150,66 +140,69 @@ void PuppiContainer::getRMSAvg(int iOpt,
...
-    //code added by Nhan, now instead for every algorithm give it all the particles
+    // code added by Nhan: now instead for every algorithm give it all the particles
     for (int i1 = 0; i1 < fNAlgos; i1++) {
-      pAlgo = fPuppiAlgo[i1].algoId(iOpt);
-      pCharged = fPuppiAlgo[i1].isCharged(iOpt);
-      pCone = fPuppiAlgo[i1].coneSize(iOpt);
-      double curVal = -1;
+      // skip cands outside of algo's etaMaxExtrap, as they would anyway be ignored inside PuppiAlgo::add (see end of the block)
+      if (not((std::abs(iConstits[i0].eta) < fPuppiAlgo[i1].etaMaxExtrap()) and
+              ((iConstits[i0].id == 1) or (iConstits[i0].id == 2))))
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc:163]:
```suggestion
      if (not(std::abs(iConstits[i0].eta) < fPuppiAlgo[i1].etaMaxExtrap() and pvRelated)
```

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 6]
```diff
@@ -309,18 +302,16 @@ std::vector<double> const &PuppiContainer::puppiWeights() {
...
     }
     //Basic Cuts
-    if (pWeight * fPFParticles[i0].pt() < fPuppiAlgo[pPupId].neutralPt(fNPV) && rParticle.id == 0)
+    if (pWeight * fPFParticles[i0].pt < fPuppiAlgo[pPupId].neutralPt(fNPV) && rParticle.id == 0)
       pWeight = 0;  //threshold cut on the neutral Pt
     // Protect high pT photons (important for gamma to hadronic recoil balance)
-    if ((fPtMaxPhotons > 0) && (rParticle.pdgId == 22) && (std::abs(fPFParticles[i0].eta()) < fEtaMaxPhotons) &&
-        (fPFParticles[i0].pt() > fPtMaxPhotons))
+    if ((fPtMaxPhotons > 0) && (rParticle.pdgId == 22) && (std::abs(fPFParticles[i0].eta) < fEtaMaxPhotons) &&
+        (fPFParticles[i0].pt > fPtMaxPhotons))
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc:309]:
```suggestion
    if (fPtMaxPhotons > 0 && rParticle.pdgId == 22 && std::abs(fPFParticles[i0].eta) < fEtaMaxPhotons &&
        fPFParticles[i0].pt > fPtMaxPhotons)
```
IMHO there is more danger to misplace a parenthesis than to confuse that ```&&```  applies after `<, >, ==`

**Review by slava77 — COMMENTED** [review]
kudos for this speedup.

**Comment by santocch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31164	e0f02d3 
- code changes are in line with the PR description and the follow up review
    - the last commits removed dependence on fastjet, but introduced a numerical difference https://github.com/cms-sw/cmssw/pull/31164/files/0c3b885fbbd8e493f396d89bcdc0687cf52f0818..e0f02d36a5b2d0fd5c5fbe53358cbb902f1700cf in the way how rapidity/pseudorapidity are set, something to be noted/bypassed in a possible backport
- jenkins tests pass and comparisons with the baseline show small differences in puppi-related variables, consistent with the change in the numerical precision of the kinematics
-  

curiously enough, there was no actual overlap in the context lines with #31174

**Comment by missirol** [comment]
> the last commits removed dependence on fastjet, but introduced a numerical difference

@slava77 

In case it helps: numerical changes in the outputs are introduced only by the very last commit of the PR, i.e. e0f02d36a5b2
(we didn't verify this explicitly in the PR review, but I had checked it locally).

**Comment by silviodonato** [comment]
+1

**Comment by missirol** [comment]
@slava77 , one late question from my side: would it be okay to backport this (except for the last commit) to `11_1_X`, so that it can be used for the Phase-2 HLT TDR studies?

**Comment by slava77** [comment]
> @slava77 , one late question from my side: would it be okay to backport this (except for the last commit) to `11_1_X`, so that it can be used for the Phase-2 HLT TDR studies?

sure, but please limit it to the commit range that does not change the results.

**Comment by kpedro88** [comment]
@missirol thanks for this PR! It's a worthy followup to my initial effort in #23270.

Could you make the same limited backport to 10_6_X? It would be nice to propagate this speedup to users for ultra-legacy analysis.

**Comment by missirol** [comment]
> Could you make the same limited backport to 10_6_X? It would be nice to propagate this speedup to users for ultra-legacy analysis.

I'm looking into a backport of this PR to `10_6_X` (taking into account the no-change policy). It would not be entirely trivial, because the Puppi implementation in earlier releases (incl. `10_6`) is significantly different; nevertheless, I'll give it a try.

**Comment by slava77** [comment]
> I'm looking into a backport of this PR to `10_6_X` (taking into account the no-change policy). It would not be entirely trivial, because the Puppi implementation in earlier releases (incl. `10_6`) is significantly different; nevertheless, I'll give it a try.

Thank you.

**Comment by missirol** [comment]
Backport to `10_6_X` tried in #31290.

**Pull request merged**